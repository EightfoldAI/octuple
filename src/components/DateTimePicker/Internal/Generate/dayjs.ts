import type { Dayjs } from 'dayjs';
import dayjs from 'dayjs';
import weekday from 'dayjs/plugin/weekday';
import localeData from 'dayjs/plugin/localeData';
import weekOfYear from 'dayjs/plugin/weekOfYear';
import weekYear from 'dayjs/plugin/weekYear';
import advancedFormat from 'dayjs/plugin/advancedFormat';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import type { GenerateConfig } from '.';

dayjs.extend(customParseFormat);
dayjs.extend(advancedFormat);
dayjs.extend(weekday);
dayjs.extend(localeData);
dayjs.extend(weekOfYear);
dayjs.extend(weekYear);

dayjs.extend((_o, c) => {
    // todo support Wo (ISO week)
    const proto = c.prototype;
    const oldFormat = proto.format;
    proto.format = function f(formatStr: string) {
        const str = (formatStr || '').replace('Wo', 'wo');
        return oldFormat.bind(this)(str);
    };
});

// The below mappings should map to langs currently supported by Eightfold
// Dayjs currently supported langs: https://github.com/iamkun/dayjs/tree/dev/src/locale
type IlocaleMapObject = Record<string, string>;
const localeMap: IlocaleMapObject = {
    cs_CZ: 'cs', // čeština
    da_DK: 'da', // Dansk
    de_DE: 'de', // Deutsch
    el_GR: 'el', // Ελληνικά
    en_GB: 'en-gb', // English (United Kingdom)
    en_US: 'en', // English (United States)
    es_ES: 'es', // Español
    es_DO: 'es-do', // Español (Dominican Republic)
    es_MX: 'es-mx', // Español (Mexico)
    fi_FI: 'fi', // Suomi
    fr_BE: 'fr', // Français (Belgium) TODO: dayjs has no fr_BE locale, use fr
    fr_CA: 'fr-ca', // Français (Canada)
    fr_FR: 'fr', // Français
    he_IL: 'he', // עברית
    hr_HR: 'hr', // Hrvatski
    ht_HT: 'ht', // Haitian
    hu_HU: 'hu', // Magyar
    it_IT: 'it', // Italiano
    ja_JP: 'ja', // 日本語
    ko_KR: 'ko', // 한국어
    ms_MY: 'ms', // Bahasa melayu
    nb_NO: 'nb', // Norsk
    nl_BE: 'nl-be', // Nederlands (Belgium)
    nl_NL: 'nl', // Nederlands
    pl_PL: 'pl', // Polski
    pt_BR: 'pt-br', // Português (Brazil)
    pt_PT: 'pt', // Português
    ru_RU: 'ru', // Pусский
    sv_SE: 'sv', // Svenska
    th_TH: 'th', // ภาษาไทย
    tr_TR: 'tr', // Türkçe
    uk_UA: 'uk', // Yкраїнська
    zh_CN: 'zh-cn', // 中文 (简体)
    zh_TW: 'zh-tw', // 中文 (繁體)
};

const parseLocale = (locale: string) => {
    const mapLocale = localeMap[locale];
    return mapLocale || locale.split('_')[0];
};

const generateConfig: GenerateConfig<Dayjs> = {
    // get
    getNow: () => dayjs(),
    getFixedDate: (string) => dayjs(string, ['YYYY-M-DD', 'YYYY-MM-DD']),
    getEndDate: (date) => date.endOf('month'),
    getWeekDay: (date) => {
        const clone = date.locale('en');
        return clone.weekday() + clone.localeData().firstDayOfWeek();
    },
    getYear: (date) => date.year(),
    getMonth: (date) => date.month(),
    getDate: (date) => date.date(),
    getHour: (date) => date.hour(),
    getMinute: (date) => date.minute(),
    getSecond: (date) => date.second(),

    // set
    addYear: (date, diff) => date.add(diff, 'year'),
    addMonth: (date, diff) => date.add(diff, 'month'),
    addDate: (date, diff) => date.add(diff, 'day'),
    setYear: (date, year) => date.year(year),
    setMonth: (date, month) => date.month(month),
    setDate: (date, num) => date.date(num),
    setHour: (date, hour) => date.hour(hour),
    setMinute: (date, minute) => date.minute(minute),
    setSecond: (date, second) => date.second(second),

    // Compare
    isAfter: (date1, date2) => date1.isAfter(date2),
    isValidate: (date) => date.isValid(),

    locale: {
        getWeekFirstDay: (locale) =>
            dayjs().locale(parseLocale(locale)).localeData().firstDayOfWeek(),
        getWeekFirstDate: (locale, date) =>
            date.locale(parseLocale(locale)).weekday(0),
        getWeek: (locale, date) => date.locale(parseLocale(locale)).week(),
        getShortWeekDays: (locale) =>
            dayjs().locale(parseLocale(locale)).localeData().weekdaysMin(),
        getShortMonths: (locale) =>
            dayjs().locale(parseLocale(locale)).localeData().monthsShort(),
        format: (locale, date, format) =>
            date.locale(parseLocale(locale)).format(format),
        parse: (locale, text, formats) => {
            const localeStr = parseLocale(locale);
            for (let i = 0; i < formats.length; i += 1) {
                const format = formats[i];
                const formatText = text;
                if (format.includes('wo') || format.includes('Wo')) {
                    // parse Wo
                    const year = formatText.split('-')[0];
                    const weekStr = formatText.split('-')[1];
                    const firstWeek = dayjs(year, 'YYYY')
                        .startOf('year')
                        .locale(localeStr);
                    for (let j = 0; j <= 52; j += 1) {
                        const nextWeek = firstWeek.add(j, 'week');
                        if (nextWeek.format('Wo') === weekStr) {
                            return nextWeek;
                        }
                    }
                    return null;
                }
                const date = dayjs(formatText, format).locale(localeStr);
                if (date.isValid()) {
                    return date;
                }
            }

            return null;
        },
    },
};

export default generateConfig;
