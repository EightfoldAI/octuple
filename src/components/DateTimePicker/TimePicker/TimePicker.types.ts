import type { Dayjs } from 'dayjs';
import type {
  PickerTimeProps,
  RangePickerTimeProps,
} from '../DatePicker/Generate/Generate.types';

export interface TimePickerLocale {
  placeholder?: string;
  rangePlaceholder?: [string, string];
}

export interface TimeRangePickerProps
  extends Omit<RangePickerTimeProps<Dayjs>, 'picker'> {
  popupClassNames?: string;
}

export interface TimePickerProps
  extends Omit<PickerTimeProps<Dayjs>, 'picker'> {
  addon?: () => React.ReactNode;
  popupClassNames?: string;
}
