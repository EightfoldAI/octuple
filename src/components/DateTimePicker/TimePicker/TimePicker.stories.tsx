import React, { useState } from 'react';
import dayjs from 'dayjs';
import type { Dayjs } from 'dayjs';
import { Stories } from '@storybook/addon-docs';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import TimePicker from './TimePicker';
import { DatePickerShape, DatePickerSize } from '../DatePicker';
import { Stack } from '../../Stack';
import { Button, ButtonSize, ButtonVariant } from '../../Button';

const { RangePicker } = TimePicker;

export default {
  title: 'Time Picker',
  parameters: {
    docs: {
      page: (): JSX.Element => (
        <main>
          <article>
            <section>
              <h1>Time Picker</h1>
              <p>
                Use the <b>TimePicker</b> To select or input a time. By clicking
                the input box, you can select a time from a popup.
              </p>
            </section>
            <section>
              <Stories includePrimary title="" />
            </section>
          </article>
        </main>
      ),
    },
  },
  argTypes: {
    allowClear: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    autoFocus: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    bordered: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    changeOnBlur: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    disabled: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    dropdownClassNames: {
      control: { type: 'text' },
    },
    inputReadOnly: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    open: {
      options: [true, false],
      control: { type: 'inline-radio' },
    },
    placeholder: {
      control: { type: 'text' },
    },
    shape: {
      options: [
        DatePickerShape.Rectangle,
        DatePickerShape.Pill,
        DatePickerShape.Underline,
      ],
      control: { type: 'inline-radio' },
    },
    size: {
      options: [
        DatePickerSize.Flex,
        DatePickerSize.Large,
        DatePickerSize.Medium,
        DatePickerSize.Small,
      ],
      control: { type: 'radio' },
    },
    status: {
      options: ['success', 'warning', 'error', 'validating', 'highlight', ''],
      control: 'select',
    },
    popupPlacement: {
      options: ['topLeft', 'topRight', 'bottomLeft', 'bottomRight'],
      control: { type: 'radio' },
    },
  },
} as ComponentMeta<typeof TimePicker>;

const Basic_Story: ComponentStory<typeof TimePicker> = (args) => {
  const onChange = (time: Dayjs, timeString: string) => {
    console.log(time, timeString);
  };

  return (
    <TimePicker
      {...args}
      defaultPickerValue={dayjs('00:00:00', 'HH:mm:ss')}
      onChange={onChange}
    />
  );
};

const Disabled_Story: ComponentStory<typeof TimePicker> = (args) => {
  const onChange = (time: Dayjs, timeString: string) => {
    console.log(time, timeString);
  };

  return (
    <TimePicker
      {...args}
      defaultPickerValue={dayjs('00:00:00', 'HH:mm:ss')}
      onChange={onChange}
    />
  );
};

const Readonly_Story: ComponentStory<typeof TimePicker> = (args) => {
  return (
    <TimePicker {...args} defaultPickerValue={dayjs('00:00:00', 'HH:mm:ss')} />
  );
};

const Controlled_Story: ComponentStory<typeof TimePicker> = (args) => {
  const [value, setValue] = useState<Dayjs | null>(null);

  const onChange = (time: Dayjs) => {
    setValue(time);
    console.log(time);
  };
  return <TimePicker {...args} value={value} onChange={onChange} />;
};

const Hour_and_Minute_Story: ComponentStory<typeof TimePicker> = (args) => {
  const format = 'HH:mm';
  return (
    <TimePicker
      {...args}
      defaultValue={dayjs('12:08', format)}
      format={format}
    />
  );
};

const Interval_Story: ComponentStory<typeof TimePicker> = (args) => {
  return <TimePicker {...args} />;
};

const Extra_Footer_Story: ComponentStory<typeof TimePicker> = (args) => {
  const [open, setOpen] = useState(false);

  return (
    <TimePicker
      {...args}
      open={open}
      onOpenChange={setOpen}
      renderExtraFooter={() => (
        <Button
          size={ButtonSize.Small}
          variant={ButtonVariant.Primary}
          onClick={() => setOpen(false)}
          text="OK"
        />
      )}
    />
  );
};

const Time_Range_Picker_Story: ComponentStory<typeof RangePicker> = (args) => {
  return <TimePicker.RangePicker {...args} />;
};

const Time_Range_Picker_Readonly_Story: ComponentStory<typeof RangePicker> = (
  args
) => {
  return <TimePicker.RangePicker {...args} />;
};

const Single_Borderless_Story: ComponentStory<typeof TimePicker> = (args) => {
  return (
    <Stack direction="vertical" flexGap="m">
      <TimePicker {...args} />
    </Stack>
  );
};

const Range_Borderless_Story: ComponentStory<typeof RangePicker> = (args) => {
  return (
    <Stack direction="vertical" flexGap="m">
      <RangePicker {...args} />
    </Stack>
  );
};

const Single_Status_Story: ComponentStory<typeof TimePicker> = (args) => {
  return (
    <Stack direction="vertical" flexGap="m">
      <TimePicker {...args} status="success" />
      <TimePicker {...args} status="warning" />
      <TimePicker {...args} status="error" />
      <TimePicker {...args} status="highlight" />
    </Stack>
  );
};

const Range_Status_Story: ComponentStory<typeof RangePicker> = (args) => {
  return (
    <Stack direction="vertical" flexGap="m">
      <TimePicker.RangePicker {...args} status="success" />
      <TimePicker.RangePicker {...args} status="warning" />
      <TimePicker.RangePicker {...args} status="error" />
      <TimePicker.RangePicker {...args} status="highlight" />
    </Stack>
  );
};

export const Basic = Basic_Story.bind({});
export const Readonly = Readonly_Story.bind({});
export const Disabled = Disabled_Story.bind({});
export const Controlled = Controlled_Story.bind({});
export const Hour_and_Minute = Hour_and_Minute_Story.bind({});
export const Interval = Interval_Story.bind({});
export const Extra_Footer = Extra_Footer_Story.bind({});
export const Time_Range_Picker = Time_Range_Picker_Story.bind({});
export const Time_Range_Picker_Readonly = Time_Range_Picker_Readonly_Story.bind(
  {}
);
export const Single_Borderless = Single_Borderless_Story.bind({});
export const Range_Borderless = Range_Borderless_Story.bind({});
export const Single_Status = Single_Status_Story.bind({});
export const Range_Status = Range_Status_Story.bind({});

// Storybook 6.5 using Webpack >= 5.76.0 automatically alphabetizes exports,
// this line ensures they are exported in the desired order.
// See https://www.npmjs.com/package/babel-plugin-named-exports-order
export const __namedExportsOrder = [
  'Basic',
  'Readonly',
  'Disabled',
  'Controlled',
  'Hour_and_Minute',
  'Interval',
  'Extra_Footer',
  'Time_Range_Picker',
  'Time_Range_Picker_Readonly',
  'Single_Borderless',
  'Range_Borderless',
  'Single_Status',
  'Range_Status',
];

const pickerArgs: Object = {
  classNames: 'my-picker-class',
  id: 'myPickerInputId',
  shape: DatePickerShape.Rectangle,
  size: DatePickerSize.Medium,
  changeOnBlur: true,
  nowButtonProps: null,
  okButtonProps: { variant: ButtonVariant.Primary },
  showNow: true,
  showOk: true,
};

Basic.args = {
  ...pickerArgs,
};

Readonly.args = {
  ...pickerArgs,
  readonly: true,
};

Disabled.args = {
  ...pickerArgs,
  disabled: true,
};

Controlled.args = {
  ...pickerArgs,
};

Hour_and_Minute.args = {
  ...pickerArgs,
};

Interval.args = {
  ...pickerArgs,
  minuteStep: 15,
  secondStep: 10,
};

Extra_Footer.args = {
  ...pickerArgs,
};

Time_Range_Picker.args = {
  ...pickerArgs,
};

Time_Range_Picker_Readonly.args = {
  ...pickerArgs,
  readonly: [true, true],
};

Single_Borderless.args = {
  ...pickerArgs,
  bordered: false,
};

Range_Borderless.args = {
  ...pickerArgs,
  bordered: false,
};

Single_Status.args = {
  ...pickerArgs,
};

Range_Status.args = {
  ...pickerArgs,
};
