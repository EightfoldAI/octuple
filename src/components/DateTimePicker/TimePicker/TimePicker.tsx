'use client';

import React, { forwardRef, useMemo } from 'react';
import DatePicker from '../DatePicker';
import { TimePickerProps, TimeRangePickerProps } from './TimePicker.types';

const { TimePicker: InternalTimePicker, RangePicker: InternalRangePicker } =
  DatePicker;

const RangePicker = forwardRef<any, TimeRangePickerProps>((props, ref) => (
  <InternalRangePicker
    {...props}
    dropdownClassNames={props.popupClassNames}
    picker="time"
    mode={undefined}
    ref={ref}
  />
));

const TimePicker = forwardRef<any, TimePickerProps>(
  ({ addon, renderExtraFooter, popupClassNames, ...rest }, ref) => {
    const internalRenderExtraFooter = useMemo(() => {
      if (renderExtraFooter) {
        return renderExtraFooter;
      }
      if (addon) {
        return addon;
      }
      return undefined;
    }, [addon, renderExtraFooter]);

    return (
      <InternalTimePicker
        {...rest}
        dropdownClassNames={popupClassNames}
        mode={undefined}
        ref={ref}
        renderExtraFooter={internalRenderExtraFooter}
      />
    );
  }
);

type MergedTimePicker = typeof TimePicker & {
  RangePicker: typeof RangePicker;
};

(TimePicker as MergedTimePicker).RangePicker = RangePicker;

export default TimePicker as MergedTimePicker;
