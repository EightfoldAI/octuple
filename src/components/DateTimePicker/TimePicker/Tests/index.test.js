import Enzyme, { mount } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import dayjs from 'dayjs';
import React from 'react';
import TimePicker from '../TimePicker';
import '@testing-library/jest-dom';
import MatchMediaMock from 'jest-matchmedia-mock';

Enzyme.configure({ adapter: new Adapter() });

let matchMedia;

describe('TimePicker', () => {
    beforeAll(() => {
        matchMedia = new MatchMediaMock();
    });

    afterEach(() => {
        matchMedia.clear();
    });

    it('not render clear icon when allowClear is false', () => {
        const wrapper = mount(
            <TimePicker
                defaultValue={dayjs('2000-01-01 00:00:00')}
                allowClear={false}
            />
        );
        expect(wrapper.render()).toMatchSnapshot();
    });

    // TODO: Re-enable when Octuple supports locales
    it.skip('prop locale should works', () => {
        const locale = {
            placeholder: 'Избери дата',
        };
        const wrapper = mount(
            <TimePicker
                defaultValue={dayjs('2000-01-01 00:00:00')}
                open
                locale={locale}
            />
        );
        expect(wrapper.render()).toMatchSnapshot();
    });

    // TODO Re-enable when test bug to find a component is fixed
    it.skip('should pass popupClassNames prop to Picker as dropdownClassNames prop', () => {
        const dropdownClassNames = 'myCustomClassName';
        const wrapper = mount(
            <TimePicker
                defaultOpenValue={dayjs('00:00:00', 'HH:mm:ss')}
                popupClassNames={dropdownClassNames}
            />
        );
        expect(
            wrapper.find('PickerPartial').last().prop('dropdownClassNames')
        ).toEqual(popupClassNames);
    });

    it.skip('should pass popupClassNames prop to RangePicker as dropdownClassNames prop', () => {
        const dropdownClassNames = 'myCustomClassName';
        const wrapper = mount(
            <TimePicker.RangePicker
                defaultOpenValue={dayjs('00:00:00', 'HH:mm:ss')}
                popupClassNames={dropdownClassNames}
            />
        );
        expect(
            wrapper.find('PickerPartial').last().prop('dropdownClassNames')
        ).toEqual(popupClassNames);
    });

    it('should support bordered', () => {
        const wrapper = mount(
            <TimePicker
                className="custom-class"
                defaultValue={dayjs('2000-01-01 00:00:00')}
                bordered={false}
            />
        );
        expect(wrapper.render()).toMatchSnapshot();
    });
});
