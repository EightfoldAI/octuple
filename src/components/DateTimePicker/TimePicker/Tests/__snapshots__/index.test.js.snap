version https://git-lfs.github.com/spec/v1
oid sha256:396b0276ee1e327648c038cba1c63e4176a03c0054a07de22bdeacdb6a65197a
size 41584
